File: Benchmarking/src/gen_0_21.scala
import java.lang.System.nanoTime
import scala.collection.immutable.Vector
object Gen extends App {
  case class s0 (
    p0: Vector[Vector[Double]],
    p1: Vector[Vector[Double]]
  )
  case class s1 (
    p0: Double,
    p1: s0
  )
  case class s2 (
    p0: s1,
    p1: Vector[Vector[s1]]
  )
  case class s3 (
    p0: Vector[s1],
    p1: s0
  )
  @noinline
  def f0(v0: s2, v1: Vector[s0], v2: s2, v3: Vector[s3], v4: Double): Double = {
    val v5: Double = v4 * v4
    var v15: Double = v5
    val v6: Vector[Vector[s1]] = v0.p1
    val v16: Double = v5 * v15
    val v27: Vector[s1] = v6(0)
    val v24: Double = v5 + v16
    val v14: s1 = v27(0)
    var v34: Double = v16
    val v42: Double = v5 - v24
    var v127: Double = v34
    val v102: Double = v14.p0
    val v74: Double = v102 - v42
    val v159: Double = v127 + v74
    v159
  }
  def benchmark(): Unit = {
    val v0: s2 = s2(s1(0.0, s0(Vector(Vector(1.0)), Vector(Vector(2.0)))), Vector(Vector(s1(3.0, s0(Vector(Vector(4.0)), Vector(Vector(5.0))))), Vector(s1(6.0, s0(Vector(Vector(7.0)), Vector(Vector(8.0)))))))
    val v1: Vector[s0] = Vector(s0(Vector(Vector(9.0)), Vector(Vector(10.0))), s0(Vector(Vector(11.0)), Vector(Vector(12.0))))
    val v2: s2 = s2(s1(13.0, s0(Vector(Vector(14.0)), Vector(Vector(15.0)))), Vector(Vector(s1(16.0, s0(Vector(Vector(17.0)), Vector(Vector(18.0))))), Vector(s1(19.0, s0(Vector(Vector(20.0)), Vector(Vector(21.0)))))))
    val v3: Vector[s3] = Vector(s3(Vector(s1(22.0, s0(Vector(Vector(23.0)), Vector(Vector(24.0))))), s0(Vector(Vector(25.0)), Vector(Vector(26.0)))), s3(Vector(s1(27.0, s0(Vector(Vector(28.0)), Vector(Vector(29.0))))), s0(Vector(Vector(30.0)), Vector(Vector(31.0)))), s3(Vector(s1(32.0, s0(Vector(Vector(33.0)), Vector(Vector(34.0))))), s0(Vector(Vector(35.0)), Vector(Vector(36.0)))))
    val v4: Double = 37.0
    val start = nanoTime()
    var result: Double = 37.0
    (1 to 1000).foreach { _ =>
      result = f0(v0, v1, v2, v3, v4)
    }
    val end = nanoTime()
    println(result)
    println(end - start)
  }
  benchmark()
}

File: Benchmarking/src/gen_0_21.swift
  import Dispatch
  struct s0 {
    var p0: [[Double]]
    var p1: [[Double]]
  }
  struct s1 {
    var p0: Double
    var p1: s0
  }
  struct s2 {
    var p0: s1
    var p1: [[s1]]
  }
  struct s3 {
    var p0: [s1]
    var p1: s0
  }
  @inline(never)
  func f0(_ v0: s2, _ v1: [s0], _ v2: s2, _ v3: [s3], _ v4: Double) -> Double {
    let v5: Double = v4 * v4
    var v15: Double = v5
    let v6: [[s1]] = v0.p1
    let v16: Double = v5 * v15
    let v27: [s1] = v6[0]
    let v24: Double = v5 + v16
    let v14: s1 = v27[0]
    var v34: Double = v16
    let v42: Double = v5 - v24
    var v127: Double = v34
    let v102: Double = v14.p0
    let v74: Double = v102 - v42
    let v159: Double = v127 + v74
    return v159
  }
  func benchmark() {
    let v0: s2 = s2(p0: s1(p0: 0.0, p1: s0(p0: [[1.0]], p1: [[2.0]])), p1: [[s1(p0: 3.0, p1: s0(p0: [[4.0]], p1: [[5.0]]))], [s1(p0: 6.0, p1: s0(p0: [[7.0]], p1: [[8.0]]))]])
    let v1: [s0] = [s0(p0: [[9.0]], p1: [[10.0]]), s0(p0: [[11.0]], p1: [[12.0]])]
    let v2: s2 = s2(p0: s1(p0: 13.0, p1: s0(p0: [[14.0]], p1: [[15.0]])), p1: [[s1(p0: 16.0, p1: s0(p0: [[17.0]], p1: [[18.0]]))], [s1(p0: 19.0, p1: s0(p0: [[20.0]], p1: [[21.0]]))]])
    let v3: [s3] = [s3(p0: [s1(p0: 22.0, p1: s0(p0: [[23.0]], p1: [[24.0]]))], p1: s0(p0: [[25.0]], p1: [[26.0]])), s3(p0: [s1(p0: 27.0, p1: s0(p0: [[28.0]], p1: [[29.0]]))], p1: s0(p0: [[30.0]], p1: [[31.0]])), s3(p0: [s1(p0: 32.0, p1: s0(p0: [[33.0]], p1: [[34.0]]))], p1: s0(p0: [[35.0]], p1: [[36.0]]))]
    let v4: Double = 37.0
    let start = DispatchTime.now().uptimeNanoseconds
    var result: Double = 37.0
    for _ in 1...1000 {
      result = f0(v0, v1, v2, v3, v4)
    }
    let end = DispatchTime.now().uptimeNanoseconds
    print(result)
    print(end - start)
  }
  benchmark()
